client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,319 @@
+/* eslint-disable no-shadow */
+import { Controller } from '@hotwired/stimulus';
+import { debounce } from '../utils/debounce';
+
```

Comment:
```suggestion

/**
 * Returns a promise that will resolve after either the animation, translation
 * or the max delay of time is reached.
 *
 * If maxDelay is provided as zero or a falsey value, the promise resolve immediately.
 */
const afterTransition = (element: HTMLElement, { maxDelay = 300 } = {}) => {
  /**
   * Allow the passing of an initial value to the resolved promise.
   * If nothings is passed, the event will be passed to the promise.
   */
  let initValue: any;

  const promise = new Promise<AnimationEvent | TransitionEvent | undefined>(
    (resolve) => {
      if (!maxDelay) {
        resolve(initValue);
        return;
      }

      let timer: number | undefined;

      const finish = (event: AnimationEvent | TransitionEvent | undefined) => {
        if (event && event.target !== element) return;
        window.clearTimeout(timer);
        element.removeEventListener('transitionend', finish);
        element.removeEventListener('animationend', finish);
        resolve(initValue || event);
      };

      element.addEventListener('animationend', finish);
      element.addEventListener('transitionend', finish);
      timer = window.setTimeout(finish, maxDelay);
    },
  );

  return (init: any) => {
    initValue = typeof init === 'function' ? init() : init;
    return promise;
  };
};

```

OK this is a bit confusing at first but this is a proposed better way to handle 'set some classes and then do something after the classes' animation finishes'.

It's based on the animate.css library recommendation here https://animate.style/#javascript

I will update examples below.